### PR TITLE
Make dashboard bind to IPv4 & IPv6 when available

### PR DESCRIPTION
**Pull request by benjamincburns**
## PR description

Addresses issue #5690.

This change makes it so that the dashboard server and dashboard message bus listens on both IPv4 and IPv6 for a given hostname. In the event that one of the two protocols is not available for a given hostname, it will only listen on the protocol that is available.

## Testing instructions

1. Run `truffle dashboard` in your project of choice.
2. Ensure that you can connect to it via http://127.0.0.1:24012 (note the port number may be different on your machine)
3. Ensure that you can connect to it via http://[::1]:24012 (note the port number may be different on your machine)
4. Configure the `dashboard` network to connect to host `127.0.0.1` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.
5. Configure the `dashboard` network to connect to host `::1` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.
6. Configure the `dashboard` network to connect to host `localhost` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if documentation updates are required.

## Breaking changes and new features

- [x] I have added any needed `breaking-change` and `new-feature` labels for the appropriate packages.

## Note to reviewers

There's a bit of unfortunate code duplication here. I would have preferred to put the logic for resolving the IPs of a given local hostname into the `@truffle/dashboard-message-bus-common` package, but the webpack operation in the build for the dashboard frontend choked on the node-specific modules that are used by this logic (the `os` and `net` modules). As a result, I decided it was easier to _temporarily_ duplicate this code between the `utils` module in the dashboard server, and the `utils` module in the `dashboard-message-bus` module.

That said, we're still not breaking "the rule of three" here, so hopefully this can be approved as-is. However if #5621 happens to be ready to merge before this PR, I can rebase this code on `develop`, move the logic to `@dashboard-message-bus-common`, and drop the `net` and `os` packages in the dashboard's newly-ejected webpack config.

### CODE REVIEW

**Comment by OnlyOneJMJQ** [comment]
Will give this a try today and get back to you!

**Comment by benjamincburns** [comment]
I think it's safe to review this while we're waiting on feedback from Josh, as it's a fairly small change and I've already verified by other means (`lsof -i -P | grep LISTEN`) that it has the intended result (listening on both IPv6 _and_ IPv4 when available).

**Comment by benjamincburns** [comment]
Just confirming that @OnlyOneJMJQ got back to me via Slack and confirmed that this change does indeed fix the issue he observed in #5690

File: packages/dashboard/lib/DashboardServer.ts [hunk 1]
```diff
@@ -104,15 +107,26 @@ export class DashboardServer {
...
-        if (this.autoOpen) {
-          const host = this.host === "0.0.0.0" ? "localhost" : this.host;
-          open(`http://${host}:${this.port}`);
-        }
-        resolve();
-      });
-    });
+    const bindIpAddresses = await resolveBindHostnameToAllIps(this.host);
+    this.httpServers = await Promise.all(
+      bindIpAddresses.map(async bindIpAddress => {
```
Comment (davidmurdoch) [inline packages/dashboard/lib/DashboardServer.ts:112]:
```suggestion
      bindIpAddresses.map(bindIpAddress => {
```

File: packages/dashboard-message-bus/lib/utils.ts [hunk 2]
```diff
@@ -79,3 +92,21 @@ export const sendAndAwait = (socket: WebSocket, message: Message) => {
     socket.send(encodedMessage);
   });
 };
+
+const getLocalInterfaceAddresses = () => {
+  return Object.keys(networkInterfaces()).flatMap(name =>
+    networkInterfaces()[name].map(iface => iface.address)
+  );
```
Comment (davidmurdoch) [inline packages/dashboard-message-bus/lib/utils.ts:99]: I believe `networkInterfaces` ends up in a system call, so the results can change between calls, even within a single node.js event loop. I could be wrong, but I'd unless you want to investigate i'd just cache the return of `networkInterfaces()` and use it for the keys and within the `flatMap`. (but if you do investigate and learn anything interesting please teach me your learnings!)

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: packages/dashboard-message-bus/lib/utils.ts (hunk 2 above)
Comment (benjamincburns) [reply packages/dashboard-message-bus/lib/utils.ts:99]: Ah good catch - will do.

**Comment by benjamincburns** [comment]
Argh, so this is frustrating... I added better error handling for the case when the user mistypes their desired hostname, but it turns out that the error handling code is unreachable due to [how we provision additional ports for the websocket servers](https://github.com/trufflesuite/truffle/blob/6af18fad1fae0b9c5d4316c4280b3face6fa48c0/packages/dashboard/lib/DashboardServer.ts#L156-L159). That is, when I purposefully test with a mistyped hostname, it throws `Error: getaddrinfo ENOTFOUND` from that line.

The bigger issue is that the `getPort` function that we're using can't find an unused port that's common between multiple local IP addresses. If I call it twice (once for IPv6 and once for IPv4) it may return different ports.

I think this means that in order to make this change properly I need to refactor the message bus to reuse the `http.Server` instance(s) that `DashboardServer` sets up to serve up the dashboard webpage.

cc @cliffoo as FYI

**Comment by eggplantzzz** [comment]
Closing this for now as the changes here are too big for something that we are not able to reliably reproduce on our end. If we can reproduce this in the future we'll revisit it.

**Pull request closed without merging**